chia/full_node/mempool_check_conditions.py

```diff
@@ -132,40 +134,222 @@ def mempool_assert_my_amount(condition: ConditionWithArgs, unspent: CoinRecord)
+        return ConditionCost.RESERVE_FEE.value, parse_fee(args)
+    elif condition is ConditionOpcode.CREATE_COIN_ANNOUNCEMENT:
+        return ConditionCost.CREATE_COIN_ANNOUNCEMENT.value, parse_announcement(args)
+    elif condition is ConditionOpcode.ASSERT_COIN_ANNOUNCEMENT:
+        return ConditionCost.ASSERT_COIN_ANNOUNCEMENT.value, parse_hash(args, Err.ASSERT_ANNOUNCE_CONSUMED_FAILED)
+    elif condition is ConditionOpcode.CREATE_PUZZLE_ANNOUNCEMENT:
+        return ConditionCost.CREATE_PUZZLE_ANNOUNCEMENT.value, parse_announcement(args)
+    elif condition is ConditionOpcode.ASSERT_PUZZLE_ANNOUNCEMENT:
+        return ConditionCost.ASSERT_PUZZLE_ANNOUNCEMENT.value, parse_hash(args, Err.ASSERT_ANNOUNCE_CONSUMED_FAILED)
+    elif condition is ConditionOpcode.ASSERT_MY_PARENT_ID:
+        return ConditionCost.ASSERT_MY_PARENT_ID.value, parse_coin_id(args, Err.ASSERT_MY_PARENT_ID_FAILED)
+    elif condition is ConditionOpcode.ASSERT_MY_PUZZLEHASH:
+        return ConditionCost.ASSERT_MY_PUZZLEHASH.value, parse_hash(args, Err.ASSERT_MY_PUZZLEHASH_FAILED)
+    elif condition is ConditionOpcode.ASSERT_MY_AMOUNT:
+        return ConditionCost.ASSERT_MY_AMOUNT.value, parse_amount(args)
+    else:
+        raise ValidationError(Err.INVALID_CONDITION)
+
+
+opcodes: Set[bytes] = set(item.value for item in ConditionOpcode)
+
+
+def parse_condition(cond: SExp, safe_mode: bool) -> Tuple[int, Optional[ConditionWithArgs]]:
+    condition = cond.first().as_atom()
+    if condition in opcodes:
+        opcode: ConditionOpcode = ConditionOpcode(condition)
+        cost, args = parse_condition_args(cond.rest(), opcode)
+        cvl = ConditionWithArgs(opcode, args) if args is not None else None
+    elif not safe_mode:
+        opcode = ConditionOpcode.UNKNOWN
+        cvl = ConditionWithArgs(opcode, cond.rest().as_atom_list())
+        cost = 0
+    else:
+        raise ValidationError(Err.INVALID_CONDITION)
+    return cost, cvl
+
+
+def get_name_puzzle_conditions(
+    generator: BlockGenerator, max_cost: int, *, cost_per_byte: int, safe_mode: bool
+) -> NPCResult:
+    """
+    This executes the generator program and returns the coins and their
+    conditions. If the cost of the program (size, CLVM execution and conditions)
+    exceed max_cost, the function fails. In order to accurately take the size
+    of the program into account when calculating cost, cost_per_byte must be
+    specified.
+    safe_mode determines whether the clvm program and conditions are executed in
+    strict mode or not. When in safe/strict mode, unknow operations or conditions
+    are considered failures. This is the mode when accepting transactions into
+    the mempool.
+    """
     try:
         block_program, block_program_args = setup_generator_args(generator)
+        max_cost -= len(bytes(generator.program)) * cost_per_byte
+        if max_cost < 0:
+            return NPCResult(uint16(Err.INVALID_BLOCK_COST.value), [], uint64(0))
```

Comment:
this function specifies a tighter max cost now. Not just by taking the size-cost into account, but also by taking the conditions cost into account, as they are parsed.